File: wobblesoccer/core/ai.py
"""Lightweight scripted AI.

Produces a raw action vector (see :mod:`action`) for every player.  The
simulation then overwrites the one row belonging to the agent/human, so AI and
learned policies use exactly the same control channel.

The behaviour is intentionally simple and readable: a keeper that hugs its line,
a chaser that goes for the ball, the rest holding a ball-reactive formation, and
basic dribble / pass / shoot decisions for whoever is on the ball.
"""

from __future__ import annotations

import numpy as np

from . import config as C


def base_formation(team_size: int) -> np.ndarray:
    """Home positions for team 0 as fractions of (HALF_LENGTH, HALF_WIDTH).

    Index 0 is the keeper.  Team 0 defends -x, so homes sit in the -x half.
    """
    homes = np.zeros((team_size, 2), dtype=np.float64)
    homes[0] = (-0.9, 0.0)                      # keeper
    outfield = team_size - 1
    back = outfield // 2
    front = outfield - back

    def line(n, x):
        zs = [0.0] if n <= 1 else list(np.linspace(-0.55, 0.55, n))
        return [(x, z) for z in zs]

    homes[1:1 + back] = line(back, -0.55)
    homes[1 + back:] = line(front, -0.12)
    return homes


def formation_world(team_size: int, team: int) -> np.ndarray:
    """Base home positions in world x/z for a given team."""
    homes = base_formation(team_size)
    world = np.empty_like(homes)
    sign = 1.0 if team == 0 else -1.0
    world[:, 0] = sign * homes[:, 0] * C.HALF_LENGTH
    world[:, 1] = homes[:, 1] * C.HALF_WIDTH
    return world


def _goal_center(team: int) -> np.ndarray:
    """The goal a given team is attacking."""
    x = C.HALF_LENGTH if team == 0 else -C.HALF_LENGTH
    return np.array([x, 0.0])


def _keeper_index(team: int, team_size: int) -> int:
    return 0 if team == 0 else team_size


def compute_actions(state, rng) -> np.ndarray:
    """Return an (N, 6) array of raw actions, one per player."""
    N = state.num_players
    ts = state.team_size
    actions = np.zeros((N, 6), dtype=np.float32)
    actions[:, 4] = -1.0  # pass off
    actions[:, 5] = -1.0  # shoot off

    pos = state.player_pos[:, [0, 2]]            # (N, 2) x/z
    ball = state.ball_pos[[0, 2]]
    ball_vel = state.ball_vel[[0, 2]]
    intercept = ball + ball_vel * C.CHASE_LEAD

    # nearest chaser per team (outfield preferred but keeper can join if closest)
    dist_to_ball = np.linalg.norm(pos - ball, axis=1)
    chaser = {}
    for t in (0, 1):
        idx = state.team_indices(t)
        chaser[t] = int(idx[np.argmin(dist_to_ball[idx])])

    formations = {t: formation_world(ts, t) for t in (0, 1)}
    ball_x_norm = float(np.clip(state.ball_pos[0] / C.HALF_LENGTH, -1.0, 1.0))

    for i in range(N):
        t = int(state.team[i])
        opp = 1 - t
        goal = _goal_center(t)
        is_keeper = (i == _keeper_index(t, ts))

        # --- keeper -------------------------------------------------------
        if is_keeper:
            own_goal_x = -C.HALF_LENGTH if t == 0 else C.HALF_LENGTH
            tgt = np.array([own_goal_x * 0.93,
                            float(np.clip(ball[1] * 0.6,
                                          -C.GOAL_HALF_WIDTH, C.GOAL_HALF_WIDTH))])
            if state.possession == i:
                # clear it up the pitch toward our furthest-forward team-mate
                mates = [j for j in state.team_indices(t) if j != i]
                fwd = max(mates, key=lambda j: (1 if t == 0 else -1) * pos[j, 0])
                aim = _aim_to(pos[i], pos[fwd], rng)
                actions[i, 2:4] = aim
                actions[i, 4] = 1.0
            else:
                actions[i, 0:2] = _move_to(pos[i], tgt)
            continue

        # --- on the ball --------------------------------------------------
        if state.possession == i:
            d_goal = float(np.linalg.norm(pos[i] - goal))
            near_opp = _nearest_opponent_dist(pos, state.team, i, opp)
            if d_goal < C.SHOOT_RANGE:
                aim = _aim_to(pos[i], goal, rng)
                actions[i, 2:4] = aim * float(np.clip(d_goal / C.SHOOT_RANGE, 0.55, 1.0))
                actions[i, 5] = 1.0
            else:
                mate = _best_pass_target(state, pos, i, t, opp, goal)
                if mate is not None and near_opp < 2.6:
                    aim = _aim_to(pos[i], pos[mate], rng)
                    d = float(np.linalg.norm(pos[i] - pos[mate]))
                    actions[i, 2:4] = aim * float(np.clip(d / 18.0, 0.4, 1.0))
                    actions[i, 4] = 1.0
                else:
                    # dribble toward goal
                    actions[i, 0:2] = _move_to(pos[i], goal, jitter=rng.normal(0, 0.05, 2))
                    actions[i, 2:4] = _aim_to(pos[i], goal, rng) * 0.2
            continue

        # --- off the ball -------------------------------------------------
        if i == chaser[t]:
            actions[i, 0:2] = _move_to(pos[i], intercept, jitter=rng.normal(0, 0.04, 2))
        else:
            home = formations[t][_home_slot(state, i, t, ts)]
            # push the line up/down the pitch with the ball
            home = home.copy()
            home[0] += ball_x_norm * 0.45 * C.HALF_LENGTH * (1 if t == 0 else 1)
            home[0] = float(np.clip(home[0], -C.HALF_LENGTH * 0.97, C.HALF_LENGTH * 0.97))
            actions[i, 0:2] = _move_to(pos[i], home, slow=True)

    return actions


# --- small helpers --------------------------------------------------------

def _home_slot(state, i, t, ts) -> int:
    """Map a player index to its formation row (0..ts-1)."""
    return i if t == 0 else i - ts


def _move_to(src, dst, jitter=None, slow=False):
    d = dst - src
    n = float(np.hypot(d[0], d[1]))
    if n < 1e-6:
        return np.zeros(2, dtype=np.float32)
    v = d / n
    if jitter is not None:
        v = v + jitter
    # ease off when nearly there so players settle instead of vibrating
    scale = min(1.0, n / 1.5) if slow else 1.0
    out = v * scale
    on = float(np.hypot(out[0], out[1]))
    if on > 1.0:
        out /= on
    return out.astype(np.float32)


def _aim_to(src, dst, rng):
    d = dst - src
    n = float(np.hypot(d[0], d[1]))
    if n < 1e-6:
        return np.array([1.0, 0.0], dtype=np.float32)
    ang = np.arctan2(d[1], d[0]) + rng.normal(0, C.AIM_NOISE)
    return np.array([np.cos(ang), np.sin(ang)], dtype=np.float32)


def _nearest_opponent_dist(pos, team, i, opp):
    idx = np.where(team == opp)[0]
    if len(idx) == 0:
        return 1e9
    return float(np.min(np.linalg.norm(pos[idx] - pos[i], axis=1)))


def _best_pass_target(state, pos, i, t, opp, goal):
    """Pick a team-mate that is more advanced and reasonably open, else None."""
    mates = [j for j in state.team_indices(t) if j != i]
    best, best_score = None, -1e9
    for j in mates:
        advance = -float(np.linalg.norm(pos[j] - goal))        # closer to goal is better
        openness = _nearest_opponent_dist(pos, state.team, j, opp)
        ahead = (pos[j, 0] - pos[i, 0]) * (1 if t == 0 else -1)
        if ahead <= 1.0:
            continue
        score = advance + 0.8 * openness
        if score > best_score:
            best, best_score = j, score
    return best

File: web/src/sim/ai.ts
/** Scripted teammates & opponents — ported from `wobblesoccer/core/ai.py`. */
import { C } from "./config";
import type { RNG } from "./rng";
import type { State, Vec2 } from "./state";
import type { Action } from "./action";

export function baseFormation(teamSize: number): number[][] {
  const homes: number[][] = [[-0.9, 0.0]]; // keeper
  const outfield = teamSize - 1;
  const back = Math.floor(outfield / 2);
  const front = outfield - back;
  const line = (n: number, x: number): number[][] => {
    if (n <= 1) return [[x, 0.0]];
    const out: number[][] = [];
    for (let i = 0; i < n; i++) out.push([x, -0.55 + (1.1 * i) / (n - 1)]);
    return out;
  };
  return homes.concat(line(back, -0.55), line(front, -0.12));
}

export function formationWorld(teamSize: number, team: number): number[][] {
  const sign = team === 0 ? 1 : -1;
  return baseFormation(teamSize).map((h) => [sign * h[0] * C.HALF_LENGTH, h[1] * C.HALF_WIDTH]);
}

function goalCenter(team: number): Vec2 {
  return [team === 0 ? C.HALF_LENGTH : -C.HALF_LENGTH, 0];
}

export function keeperIndex(team: number, teamSize: number): number {
  return team === 0 ? 0 : teamSize;
}

function dist(a: Vec2, b: Vec2): number {
  return Math.hypot(a[0] - b[0], a[1] - b[1]);
}

function moveTo(src: Vec2, dst: Vec2, jitter?: Vec2, slow = false): Vec2 {
  const d: Vec2 = [dst[0] - src[0], dst[1] - src[1]];
  const n = Math.hypot(d[0], d[1]);
  if (n < 1e-6) return [0, 0];
  let v: Vec2 = [d[0] / n, d[1] / n];
  if (jitter) v = [v[0] + jitter[0], v[1] + jitter[1]];
  const scale = slow ? Math.min(1, n / 1.5) : 1;
  let out: Vec2 = [v[0] * scale, v[1] * scale];
  const on = Math.hypot(out[0], out[1]);
  if (on > 1) out = [out[0] / on, out[1] / on];
  return out;
}

function aimTo(src: Vec2, dst: Vec2, rng: RNG): Vec2 {
  const d: Vec2 = [dst[0] - src[0], dst[1] - src[1]];
  const n = Math.hypot(d[0], d[1]);
  if (n < 1e-6) return [1, 0];
  const ang = Math.atan2(d[1], d[0]) + rng.normal(0, C.AIM_NOISE);
  return [Math.cos(ang), Math.sin(ang)];
}

function nearestOpponentDist(pos: Vec2[], teams: number[], i: number, opp: number): number {
  let best = 1e9;
  for (let j = 0; j < pos.length; j++) {
    if (teams[j] === opp) best = Math.min(best, dist(pos[j], pos[i]));
  }
  return best;
}

function bestPassTarget(s: State, pos: Vec2[], i: number, t: number, opp: number, goal: Vec2): number | null {
  let best: number | null = null;
  let bestScore = -1e9;
  for (const j of s.teamIndices(t)) {
    if (j === i) continue;
    const ahead = (pos[j][0] - pos[i][0]) * (t === 0 ? 1 : -1);
    if (ahead <= 1.0) continue;
    const advance = -dist(pos[j], goal);
    const openness = nearestOpponentDist(pos, s.players.map((p) => p.team), j, opp);
    const score = advance + 0.8 * openness;
    if (score > bestScore) {
      best = j;
      bestScore = score;
    }
  }
  return best;
}

export function computeActions(s: State, rng: RNG): Action[] {
  const N = s.players.length;
  const ts = s.teamSize;
  const actions: Action[] = [];
  const pos: Vec2[] = s.players.map((p) => [p.pos[0], p.pos[2]]);
  const teams = s.players.map((p) => p.team);
  const ball: Vec2 = [s.ballPos[0], s.ballPos[2]];
  const ballVel: Vec2 = [s.ballVel[0], s.ballVel[2]];
  const intercept: Vec2 = [ball[0] + ballVel[0] * C.CHASE_LEAD, ball[1] + ballVel[1] * C.CHASE_LEAD];

  const chaser: Record<number, number> = {};
  for (const t of [0, 1]) {
    const idx = s.teamIndices(t);
    let bi = idx[0];
    let bd = 1e9;
    for (const j of idx) {
      const d = dist(pos[j], ball);
      if (d < bd) {
        bd = d;
        bi = j;
      }
    }
    chaser[t] = bi;
  }

  const formations: Record<number, number[][]> = {
    0: formationWorld(ts, 0),
    1: formationWorld(ts, 1),
  };
  const ballXNorm = Math.max(-1, Math.min(1, s.ballPos[0] / C.HALF_LENGTH));

  for (let i = 0; i < N; i++) {
    const a: Action = [0, 0, 0, 0, -1, -1];
    const t = teams[i];
    const opp = 1 - t;
    const goal = goalCenter(t);
    const isKeeper = i === keeperIndex(t, ts);

    if (isKeeper) {
      const ownGoalX = t === 0 ? -C.HALF_LENGTH : C.HALF_LENGTH;
      const inBox = t === 0 ? ball[0] < -C.HALF_LENGTH * 0.55 : ball[0] > C.HALF_LENGTH * 0.55;
      if (s.possession === i) {
        // clear it up the pitch to the furthest-forward team-mate
        const mates = s.teamIndices(t).filter((j) => j !== i);
        let fwd = mates[0];
        for (const j of mates) if ((t === 0 ? 1 : -1) * pos[j][0] > (t === 0 ? 1 : -1) * pos[fwd][0]) fwd = j;
        const aim = aimTo(pos[i], pos[fwd], rng);
        a[2] = aim[0];
        a[3] = aim[1];
        a[4] = 1;
      } else if (inBox && dist(pos[i], ball) < 4.5 && s.possession < 0) {
        // rush off the line to smother a loose ball in the box
        const m = moveTo(pos[i], ball);
        a[0] = m[0];
        a[1] = m[1];
      } else {
        // hold the line and cover the angle to the ball
        const tgt: Vec2 = [
          ownGoalX * 0.92,
          Math.max(-C.GOAL_HALF_WIDTH - 0.5, Math.min(C.GOAL_HALF_WIDTH + 0.5, ball[1] * 0.85)),
        ];
        const m = moveTo(pos[i], tgt);
        a[0] = m[0];
        a[1] = m[1];
      }
      actions.push(a);
      continue;
    }

    if (s.possession === i) {
      const dGoal = dist(pos[i], goal);
      const nearOpp = nearestOpponentDist(pos, teams, i, opp);
      if (dGoal < C.SHOOT_RANGE) {
        // aim for the open corner, away from where the keeper is standing
        const gk = keeperIndex(opp, ts);
        const kz = pos[gk] ? pos[gk][1] : 0;
        const target: Vec2 = [goal[0], (kz >= 0 ? -1 : 1) * C.GOAL_HALF_WIDTH * 0.72];
        const aim = aimTo(pos[i], target, rng);
        const p = Math.max(0.7, Math.min(1, dGoal / C.SHOOT_RANGE));
        a[2] = aim[0] * p;
        a[3] = aim[1] * p;
        a[5] = 1;
      } else {
        const mate = bestPassTarget(s, pos, i, t, opp, goal);
        if (mate !== null && nearOpp < 2.6) {
          const aim = aimTo(pos[i], pos[mate], rng);
          const d = dist(pos[i], pos[mate]);
          const p = Math.max(0.4, Math.min(1, d / 18));
          a[2] = aim[0] * p;
          a[3] = aim[1] * p;
          a[4] = 1;
        } else {
          const m = moveTo(pos[i], goal, [rng.normal(0, 0.05), rng.normal(0, 0.05)]);
          a[0] = m[0];
          a[1] = m[1];
          const aim = aimTo(pos[i], goal, rng);
          a[2] = aim[0] * 0.2;
          a[3] = aim[1] * 0.2;
        }
      }
      actions.push(a);
      continue;
    }

    if (i === chaser[t]) {
      const m = moveTo(pos[i], intercept, [rng.normal(0, 0.04), rng.normal(0, 0.04)]);
      a[0] = m[0];
      a[1] = m[1];
    } else {
      const slot = t === 0 ? i : i - ts;
      const home: Vec2 = [formations[t][slot][0], formations[t][slot][1]];
      home[0] += ballXNorm * 0.45 * C.HALF_LENGTH;
      home[0] = Math.max(-C.HALF_LENGTH * 0.97, Math.min(C.HALF_LENGTH * 0.97, home[0]));
      const m = moveTo(pos[i], home, undefined, true);
      a[0] = m[0];
      a[1] = m[1];
    }
    actions.push(a);
  }
  return actions;
}
